x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/aggs/mapreduce/AbstractMapReducer.java

```diff
@@ -0,0 +1,256 @@
+     * (By design MapReduceAggregator and InternalMapReduceAggregation are generic and don't know the internal type)
+     */
+
+    @Override
+    public void writeTo(StreamOutput out) throws IOException {
+        out.writeString(aggregationName);
+    }
+
+    Writeable mapInit(BigArrays bigArrays) {
+        return doMapInit(bigArrays);
+    }
+
+    Writeable map(Stream<Tuple<String, List<Object>>> keyValues, Writeable mapReduceContext) {
+        @SuppressWarnings("unchecked")
+        MapReduceContext context = (MapReduceContext) mapReduceContext;
+
+        return doMap(keyValues, context);
+    }
+
+    Writeable mapFinalize(Writeable mapReduceContext) {
+        @SuppressWarnings("unchecked")
+        MapReduceContext context = (MapReduceContext) mapReduceContext;
+
+        return doMapFinalize(context);
+    };
+
+    Writeable reduceInit(BigArrays bigArrays) {
+        return doReduceInit(bigArrays);
+    }
+
+    Writeable reduce(Stream<Writeable> partitions, Writeable mapReduceContext) {
+        @SuppressWarnings("unchecked")
+        MapReduceContext context = (MapReduceContext) mapReduceContext;
+
+        @SuppressWarnings("unchecked")
+        Stream<MapReduceContext> p = (Stream<MapReduceContext>) partitions;
+
+        return doReduce(p, context);
+    }
+
+    Writeable combine(Stream<Writeable> partitions, Writeable mapReduceContext) {
+        @SuppressWarnings("unchecked")
+        MapReduceContext context = (MapReduceContext) mapReduceContext;
+
+        @SuppressWarnings("unchecked")
+        Stream<MapReduceContext> p = (Stream<MapReduceContext>) partitions;
+
+        return doCombine(p, context);
+    }
+
+    Writeable reduceFinalize(Writeable mapReduceContext) throws IOException {
+        @SuppressWarnings("unchecked")
+        MapReduceContext context = (MapReduceContext) mapReduceContext;
+
+        return doReduceFinalize(context);
+    };
+
+    Writeable finalizeSampling(SamplingContext samplingContext, Writeable mapReduceResult) {
+        @SuppressWarnings("unchecked")
+        MapReduceResult result = (MapReduceResult) mapReduceResult;
+
+        return doFinalizeSampling(samplingContext, result);
+    }
+
+    void writeContext(StreamOutput out, Writeable Writeable) throws IOException {
+        Writeable.writeTo(out);
+    }
+
+    MapReduceContext readMapContext(StreamInput in, BigArrays bigArrays) throws IOException {
+        // TODO: what if this isn't a map context but the result object?
+        // it seems to dangerous to assume that deserialization only happens on partial reduced objects...
+        return doReadMapContext(in, bigArrays);
+    }
+
+    // we know that MapReduceResult implements ToXContent
+    @SuppressWarnings("unchecked")
+    XContentBuilder toXContent(Writeable mapReduceResult, XContentBuilder builder, DelegatingMapParams delegatingMapParams)
+        throws IOException {
+        return ((MapReduceResult) mapReduceResult).toXContent(builder, delegatingMapParams);
+    };
```

Comment:
Just to reiterate, all these `@SuppressWarnings("unchecked")` and casting `Writeable` is a really bad code smell. Something is wrong here if we need to do this all the time for almost every method.